near-sdk/src/private/abi.rs

```diff
@@ -0,0 +1,183 @@
+use schemars::schema::{RootSchema, Schema};
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+
+/// Current version of the ABI schema format.
+const ABI_SCHEMA_SEMVER: &str = "0.1.0";
+
+/// Contract ABI.
+#[derive(Clone, Serialize, Deserialize, Debug, PartialEq)]
+pub struct AbiRoot {
+    /// Semver of the ABI schema format.
+    pub abi_schema_version: String,
+    /// Metadata information about the contract.
+    pub metadata: AbiMetadata,
+    /// Core ABI information (functions and types).
+    pub abi: Abi,
+}
+
+impl AbiRoot {
+    pub fn new(abi: Abi) -> Self {
+        Self {
+            abi_schema_version: ABI_SCHEMA_SEMVER.to_string(),
+            metadata: Default::default(),
+            abi,
+        }
+    }
+
+    pub fn combine<I: IntoIterator<Item = AbiRoot>>(abi_roots: I) -> AbiRoot {
+        let mut abi_schema_version = "".to_string();
+        let mut abis = Vec::<Abi>::new();
+
+        for abi_root in abi_roots {
+            // Set common schema version to the current schema version. Only happens with the
+            // first element.
+            if abi_schema_version.is_empty() {
+                abi_schema_version = abi_root.abi_schema_version.clone();
+            }
+
+            // Check that all ABIs conform to the same version
+            if abi_root.abi_schema_version != abi_schema_version {
+                panic!(
+                    "Conflicting ABI schema versions: {} and {}",
+                    &abi_root.abi_schema_version, abi_schema_version
+                );
+            }
+
+            // Check that all metadata is empty.
+            if !abi_root.metadata.name.is_none()
+                || !abi_root.metadata.version.is_none()
+                || !abi_root.metadata.authors.is_empty()
+                || !abi_root.metadata.other.is_empty()
+            {
+                panic!("Non-empty metadata: {:?}", &abi_root.metadata);
+            }
+
+            abis.push(abi_root.abi);
+        }
+
+        let abi = Abi::combine(abis);
+        AbiRoot { abi_schema_version, metadata: AbiMetadata::default(), abi }
+    }
+}
+
+#[derive(Clone, Serialize, Deserialize, Debug, PartialEq, Default)]
+pub struct AbiMetadata {
+    /// The name of the smart contract.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[serde(default)]
+    pub name: Option<String>,
+    /// The version of the smart contract.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[serde(default)]
+    pub version: Option<String>,
+    /// The authors of the smart contract.
+    #[serde(skip_serializing_if = "Vec::is_empty")]
+    #[serde(default)]
+    pub authors: Vec<String>,
+    /// Other arbitrary metadata.
+    #[serde(skip_serializing_if = "HashMap::is_empty")]
+    #[serde(default)]
+    #[serde(flatten)]
+    pub other: HashMap<String, String>,
```

Comment:
```suggestion
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    /// The version of the smart contract.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub version: Option<String>,
    /// The authors of the smart contract.
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub authors: Vec<String>,
    /// Other arbitrary metadata.
    #[serde(default, flatten, skip_serializing_if = "HashMap::is_empty")]
    pub other: HashMap<String, String>,
```

Optional: [You can keep these within one line](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=bf440278b3f491c556cc1d506721f372) if you want.